Juan Esteban Camacho Castro, Juan Diego Serpa Ruiz

## Importamos librerias

In [1]:
import pandas as pd


## Importamos dataset

In [11]:
df = pd.read_csv("NovaMarket_datos_crudos.csv")

## Revisamos inconsistencias en los datos

In [12]:
for col in ["ciudad_tienda", "categoria_producto", "canal_compra", "metodo_pago"]: print(col, sorted(df[col].dropna().unique()))

ciudad_tienda [' Barranquilla', ' Bogotá', ' Bucaramanga', ' Cali', ' Cartagena', ' Cúcuta', ' Ibagué', ' Manizales', ' Medellín', ' Pereira', ' Santa Marta', ' Villavicencio', 'B/manga', 'B/quilla', 'BAQ', 'BARRANQUILLA', 'BGA', 'BOG', 'BOGOTA', 'BOGOTA DC', 'BOGOTÁ', 'BUCARAMANGA', 'Barranquilla', 'Barranquilla ', 'Bogota', 'Bogota D.C.', 'Bogotá', 'Bogotá ', 'Bucaramanga', 'Bucaramanga ', 'CALI', 'CALI ', 'CARTAGENA', 'CTG', 'CUCUTA', 'Cali', 'Cali ', 'Cartagena', 'Cartagena ', 'Cartagena de Indias', 'Cucuta', 'CÚCUTA', 'Cúcuta', 'Cúcuta ', 'IBAGUE', 'IBAGUÉ', 'Ibague', 'Ibagué', 'Ibagué ', 'MANIZALES', 'MDE', 'MEDELLIN', 'MEDELLÍN', 'MZL', 'Manizales', 'Manizales ', 'Manizales, Caldas', 'Medellin', 'Medellin, Ant.', 'Medellín', 'Medellín ', 'PEREIRA', 'Pereira', 'Pereira ', 'SANTA MARTA', 'SMR', 'Santa Marta', 'Santa Marta ', 'SantaMarta', 'Santiago de Cali', 'Sta. Marta', 'VILLAVICENCIO', 'Villavicencio', 'Villavicencio ', 'barranquilla', 'bogota', 'bogota d.c.', 'bucaramanga', 'c

In [13]:
df[col].value_counts()

metodo_pago
Transferencia     127962
Tarjeta           127631
PayPal            127583
Efectivo          127452
paypal             13145
efectivo           12955
transferencia      12932
tarjeta            12807
 Efectivo           6593
Tarjeta             6581
 Transferencia      6544
Transferencia       6529
 Tarjeta            6514
EFECTIVO            6513
PAYPAL              6416
Efectivo            6406
PayPal              6393
TARJETA             6385
 PayPal             6375
TRANSFERENCIA       6284
Name: count, dtype: int64

In [14]:
df.shape[0]

640000

## Quitamos espacios y capitalizamos las palabras

In [15]:
df["ciudad_tienda"] = df["ciudad_tienda"].str.strip().str.title()

In [16]:
df["categoria_producto"] = df["categoria_producto"].str.strip().str.title()

In [17]:
df["canal_compra"] = df["canal_compra"].str.strip().str.title()

In [18]:
df["metodo_pago"] = df["metodo_pago"].str.strip().str.title()

## Volvemos a revisar la integridad del dataset y volvemos a corroborar si hay mucha variación de nombres en los datos

In [19]:
for col in ["ciudad_tienda", "categoria_producto", "canal_compra", "metodo_pago"]: print(col, sorted(df[col].dropna().unique()))

ciudad_tienda ['B/Manga', 'B/Quilla', 'Baq', 'Barranquilla', 'Bga', 'Bog', 'Bogota', 'Bogota D.C.', 'Bogota Dc', 'Bogotá', 'Bucaramanga', 'Cali', 'Cartagena', 'Cartagena De Indias', 'Ctg', 'Cucuta', 'Cúcuta', 'Ibague', 'Ibagué', 'Manizales', 'Manizales, Caldas', 'Mde', 'Medellin', 'Medellin, Ant.', 'Medellín', 'Mzl', 'Pereira', 'Santa Marta', 'Santamarta', 'Santiago De Cali', 'Smr', 'Sta. Marta', 'Villavicencio']
categoria_producto ['Belleza', 'Deportes', 'Electronica', 'Electrónica', 'Hogar', 'Jugueteria', 'Juguetería', 'Moda']
canal_compra ['App', 'Marketplace', 'Tienda', 'Web']
metodo_pago ['Efectivo', 'Paypal', 'Tarjeta', 'Transferencia']


In [20]:
df.shape[0]

640000

## Diccionarios de corrección

### Ciudad tienda

In [21]:
# Diccionario de correcciones para ciudad_tienda
correcciones_ciudad = {
    "B/Manga": "Bucaramanga",
    "B/Quilla": "Barranquilla",
    "Baq": "Barranquilla",
    "Bga": "Bucaramanga",
    "Bog": "Bogotá",
    "Bogota": "Bogotá",
    "Bogota D.C.": "Bogotá",
    "Bogota Dc": "Bogotá",
    "Cartagena De Indias": "Cartagena",
    "Ctg": "Cartagena",
    "Cucuta": "Cúcuta",
    "Ibague": "Ibagué",
    "Manizales, Caldas": "Manizales",
    "Mde": "Medellín",
    "Medellin": "Medellín",
    "Medellin, Ant.": "Medellín",
    "Mzl": "Manizales",
    "Santamarta": "Santa Marta",
    "Santiago De Cali": "Cali",
    "Smr": "Santa Marta",
    "Sta. Marta": "Santa Marta"
}

In [22]:
df["ciudad_tienda"] = df["ciudad_tienda"].replace(correcciones_ciudad)

In [29]:
df["ciudad_tienda"].unique()

<StringArray>
[       'Bogotá',     'Manizales',        'Cúcuta',   'Bucaramanga',
      'Medellín',       'Pereira',          'Cali',   'Santa Marta',
        'Ibagué', 'Villavicencio',     'Cartagena',  'Barranquilla',
             nan]
Length: 13, dtype: str

### Categoría producto

In [23]:
correcciones_categoria = {
    "Electronica": "Electrónica",
    "Jugueteria": "Juguetería"
}

In [24]:
df["categoria_producto"] = df["categoria_producto"].replace(correcciones_categoria)

In [32]:
df["categoria_producto"].unique()

<StringArray>
['Moda', 'Electrónica', 'Juguetería', 'Deportes', 'Hogar', 'Belleza']
Length: 6, dtype: str

### Canal compra

In [25]:
correcciones_canal = {
    # No hay variantes que requieran corrección
}

In [26]:
df["canal_compra"] = df["canal_compra"].replace(correcciones_canal)


In [31]:
df["canal_compra"].unique()

<StringArray>
['Tienda', 'Web', 'App', 'Marketplace']
Length: 4, dtype: str

### Método pago

In [27]:
correcciones_metodo_pago = {
    # No hay variantes que requieran corrección
}

In [28]:
df["metodo_pago"] = df["metodo_pago"].replace(correcciones_metodo_pago)

In [30]:
df["metodo_pago"].unique()

<StringArray>
['Efectivo', 'Paypal', 'Transferencia', 'Tarjeta']
Length: 4, dtype: str

## Diccionario de meses en español

In [33]:
meses_es_a_num = {"ene": "01", "feb": "02", "mar": "03", "abr": "04", "may": "05", "jun": "06", "jul": "07", "ago": "08", "sep": "09", "oct": "10", "nov": "11", "dic": "12"}

In [34]:
for mes_es, mes_num in meses_es_a_num.items(): df["fecha_pedido"] = df["fecha_pedido"].str.replace(f"-{mes_es}-", f"-{mes_num}-", regex=False)

In [35]:
df["fecha_pedido"] = pd.to_datetime(df["fecha_pedido"], dayfirst=True, errors="coerce")

In [36]:
df["fecha_pedido"].isnull().sum()

np.int64(223053)

Se encontraron alrededor de 223 mil fechas con valores nulos

## Revisión de duplicados

In [38]:
df.duplicated().sum()

np.int64(20000)

Se encontraron 20 mil registros duplicados en todo el dataset

### Duplicados correo cliente

In [39]:
df.duplicated(subset=["correo_cliente"], keep=False).sum()

np.int64(138688)

In [40]:
df[df.duplicated(subset=["correo_cliente"], keep=False)].sort_values("correo_cliente")

,id_pedido,id_cliente,fecha_pedido,canal_compra,metodo_pago,ciudad_tienda,codigo_postal,categoria_producto,producto,precio_unitario,unidades_vendidas,monto_compra,edad_cliente,correo_cliente,nivel_satisfaccion,nivel_lealtad,comentario_cliente,peso_pedido_kg,fecha_actualizacion_stock,fuga_cliente
281573,P200040,C51427,2025-07-16,Tienda,Tarjeta,Cali,7600111,Belleza,Kit de brochas,155300.0,4,621200.0,-5,cliente100040@hotmail.com,Bajo,Oro,NaN,10.4,2027-11-06,1
299542,P200040,C51427,2025-07-16,Tienda,Tarjeta,Cali,7600111,Belleza,Kit de brochas,155300.0,4,621200.0,-5,cliente100040@hotmail.com,Bajo,Oro,NaN,10.4,2027-11-06,1
542613,P200097,C53819,NaT,Web,Paypal,Bogotá,1100125,Moda,Camisa de lino,786900.0,2,1573800.0,35,cliente100097@hotmail.com,Bajo,Bronce,"Excelente calidad, volvería a comprar",10.6,2026-05-11,0
477342,P200097,C53819,NaT,Web,Paypal,Bogotá,1100125,Moda,Camisa de lino,786900.0,2,1573800.0,35,cliente100097@hotmail.com,Bajo,Bronce,"Excelente calidad, volvería a comprar",10.6,2026-05-11,0
557716,P200116,C152336,NaT,Web,Paypal,Manizales,1700101,Electrónica,Monitor 24 pulgadas,50000000.0,4,200000000.0,45,cliente100116@yahoo.es,Alto,Oro,Rápido y sin problemas,1112.0,2026-05-27,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
639976,P214545,C147031,NaT,Tienda,Tarjeta,Cartagena,1300101,Hogar,Set de cuchillos,155400.0,1,155400.0,60,NaN,Bajo,Platino,Muy buen precio,11.7,2027-08-26,0
639982,P231923,C27475,2025-01-20,Marketplace,Tarjeta,Ibagué,7300101,Belleza,Secador de pelo,213800.0,5,1069000.0,200,NaN,Bajo,Bronce,Llegó incompleto el pedido,12.3,2027-08-11,1
639986,P307845,C136455,NaT,Tienda,Paypal,Cartagena,1300105,Juguetería,Carro a control remoto,303700.0,1,303700.0,45,NaN,Alto,Plata,Llegó incompleto el pedido,14.4,2026-04-03,0
639988,P238956,C74804,2025-04-12,Web,Paypal,Villavicencio,5000103,Moda,Camisa de lino,470800.0,2,941600.0,20,NaN,Medio,Oro,Pésima experiencia con el empaque,10.2,2026-06-29,0


Se encontraron 138688 correos duplicados como por ejemplo: cliente100040@hotmail.com y cliente100097@hotmail.com. En donde ambos aparecen 2 veces en el dataset

Se van a conservar los correos más recientes ya que pueden ser una actualización de datos o un error del sistema

## Eliminación de duplicados

In [41]:
filas_antes = df.shape[0]

In [42]:
df = df.sort_values("fecha_pedido").drop_duplicates(subset=["correo_cliente"], keep="last")

In [43]:
df = df.drop_duplicates(keep="last")

In [44]:
filas_despues = df.shape[0]